lib/recipes-data/src/lib/extract-recipes.ts

```diff
@@ -37,16 +39,32 @@ export async function extractAllRecipesFromArticle(content: Content): Promise<Re
 }
 
 
-export function extractRecipeData(canonicalId: string, block: Block, sponsorship: Sponsorship[]): Array<RecipeReferenceWithoutChecksum | null> {
+export function extractRecipeData(content: Content, block: Block, sponsorship: Sponsorship[]): Array<RecipeReferenceWithoutChecksum | null> {
```

Comment:
If we have decided to pass whole `content` as parameter then do we need `block` as separate argument in this method call?
If i can see correctly then `block` is `content.blocks.main`?

Reply:
No - extractRecipeData is called from several places.  The first time it passes `articleBlocks.main` but on subsequent occasions it does a flatMap through `articleBlocks.body`

Reply:
Yes, it does flatmap and iterate through different blocks which was this method's original goal for,  but i think passing whole `content` and then its internal data `block` both are sounding confusing technically. Either we should be able to pass `publicationDate` as parameter to use further for `content.fields?.firstPublicationDate` to construct `recipeDates`?

Reply:
Yes, it does flatmap and iterate through different blocks which was this method's original goal for, but i think passing whole content and then its internal data block both are sounding confusing technically. 
Either we should be able to pass then `publicationDate` as parameter to use further for content.fields?.firstPublicationDate to construct recipeDates?

Reply:
I see your point, but on the other hand determining the firstPublicationDate and publicationDate before calling extractRecipeData would result in extra processing as I currently only need to find the fallback options if the main ones don't exist (whereas your suggestion would mean having to calculate the fallback options whether they're needed or not).  So pros and cons.

Reply:
Hmmmmmm, I see both points. I can imagine one way of doing it that's a bit different, more scala-y, that might work.

Something like this:

```typescript
interface RecipeDateTime {
  lastModified:Date;
  firstPublished:Date;
  published:Date;
}

function RecipeDateTimeFromCapi(lastModified:CapiDateTime, firstPublished:CapiDateTime, published:CapiDateTime): RecipeDateTime {
  return {
    lastModified: new Date(lastModified.dateTime.toNumber()),
    firstPublished: new Date(firstPublished.dateTime.toNumber()),
    published: new Date(published.dateTime.toNumber())
  }
}

function UpdateRecipeDateTimeFromBlock(block:Block, source:RecipeDateTime):RecipeDateTime {
  return {
    lastModified:  block.lastModifiedDate ? new Date(block.lastModifiedDate.dateTime.toNumber()) : source.lastModified,
    firstPublished: block.firstPublishedDate ? new Date(block.firstPublishedDate.dateTime.toNumber()) : source.firstPublished,
    published: block.publishedDate ? new Date(block.publishedDate.dateTime.toNumber()): source.published
  }
}
```

Then, we can call like this:

```typescript
const recipeDates = RecipeDateTimeFromCapi(block.lastModifiedDate, block.firstPublishedDate, block.publishedDate);
const getAllMainBlockRecipesIfPresent = extractRecipeData(content.id, articleBlocks.main as Block, rdt, sponsorship);

.
.
.
if (!block?.elements) return [];
  else {
   const finalRecipeDates = UpdateRecipeDateTimeFromBlock(block, recipeDates);
    return block.elements
      .filter(elem => elem.type === ElementType.RECIPE)
      .map(recp => parseJsonBlob(content.id, recp.recipeTypeData?.recipeJson as string, sponsorship, finalRecipeDates))
  }
}
```

what do you think?

Reply:
This looks really good, not confusing as well, i think we can try this if @codingWithRach also agrees.

Just one quick check: Unless I missed reading it properly Is this code taking both places `block` dates for block and source?, i think we need to pass `content`'s field's publicationdate at one place to get as source dates and then compare if block's dates are available or not otherwise contnet's dates. Let me know if not this.

Reply:
Agreed that that looks like a neater solution, although @Divs-B is correct that one of them needs to be populated from the section rather than the block.  I'll have a look at it now!

Reply:
So here's a suggestion (although it does involve calculating the fallback values whether they're required or not):

**extractAllRecipesFromArticle**

```
    const fallbackRecipeDates: RecipeDates = {
      lastModifiedDate: undefined,
      firstPublishedDate: getFallbackFirstPublishedDate(content),
      publishedDate: getFallbackPublishedDate(content)
      };
```

**extractRecipeData**

`const recipeDates: RecipeDates = updateRecipeDatesFromBlock(block, fallbackRecipeDates)`

**New functions**

```
export function getFallbackFirstPublishedDate(content: Content): Date | undefined {
  return content.fields?.firstPublicationDate ? capiDateTimeToDate(content.fields.firstPublicationDate) : undefined;
}

export function getFallbackPublishedDate(content: Content): Date | undefined {
  const feastChannel = content.channels?.find(channel => channel.channelId === 'feast');
  return feastChannel?.fields.publicationDate ? capiDateTimeToDate(feastChannel.fields.publicationDate) : undefined;
}

export function updateRecipeDatesFromBlock(block: Block, fallbackRecipeDates: RecipeDates): RecipeDates {
  return {
    lastModifiedDate: block.lastModifiedDate ? capiDateTimeToDate(block.lastModifiedDate): fallbackRecipeDates.lastModifiedDate,
    firstPublishedDate: block.firstPublishedDate ? capiDateTimeToDate(block.firstPublishedDate): fallbackRecipeDates.firstPublishedDate,
    publishedDate: block.publishedDate ? capiDateTimeToDate(block.publishedDate): fallbackRecipeDates.publishedDate,
  }
}
```

Reply:
Thanks @codingWithRach for looking into this and refining. It will be great to have refactoring in follow up PR. Let this go ahead 👍

Reply:
Yeah let's refactor it going forwards